sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/ReadTask.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+import java.io.Serializable;
+
+/**
+ * A read task returned by a data source reader and is responsible for outputting data for an RDD
+ * partition.
+ */
+public interface ReadTask<T> extends Serializable {
```

Comment:
Why not use standard interfaces like `Iterable` and `Iterator` or at least the same pattern? This interface mixes the two together by having both `open` and `next`/`get`. That makes it easy to forget to call `open` and prevents the `ReadTask` from being used more than once (unless it has been serialized to two places). I would expect `ReadTask` to be like an `Iterable` so that it can be reused if needed, and for it to return an equivalent of `Iterator` that actually does the read and tracks state. Requesting the `Iterator` from a task opens it, so you never forget.

Reply:
I agree that we should make `ReadTask` like an `Iterable`, and can return an `Iterator`. But I can't find a standard closable iterator interface in JDK, shall we create one here?

Reply:
can we just return an object with an interface like:

```java
interface Reader<T> extends Spliterator<T>, Closeable {}
```

(spliterator is a much nicer interface than iterator to implement)

or is the issue here that we can't use Java 8?

Reply:
I would use `Closeable`, which requires that multiple calls to `close` are safe. The iterator should close itself when it is exhausted, hasNext should return false, and NoSuchElementException should be thrown by next as required by Iterator.